File: grid_test.go
package main

import "testing"

func TestClone(t *testing.T) {
	g := NewGrid()
	g.Tiles[0][0] = 2
	g.Score = 10
	g2 := g.Clone()

	if !g.Equal(g2) {
		t.Fatalf("clone doesn't produce same data")
	}

	g2.Tiles[0][1] = 20
	if g.Equal(g2) {
		t.Fatalf("clones aren't independent")
	}
}

func TestMovement(t *testing.T) {
	g := NewGrid()
	g.Tiles[0][0] = 2;
	g.Tiles[0][1] = 2;
	if !g.Equal(FromSparseGrid([]int{0, 0, 2, 0, 1, 2}, 0)) {
		t.Fatalf("initialization failed")
	}

	g.Move(RIGHT)
	if !g.Equal(FromSparseGrid([]int{0, 3, 4}, 4)) {
		t.Fatalf("move right 1 failed")
	}

	g.Tiles[1][2] = 4;
	g.Move(UP)
	if !g.Equal(FromSparseGrid([]int{0, 2, 4, 0, 3, 4}, 4)) {
		t.Fatalf("move up 1 failed")
	}

	g.Move(LEFT);
	if !g.Equal(FromSparseGrid([]int{0, 0, 8}, 12)) {
		t.Fatalf("move left 1 failed")
	}

	g.Tiles[1][0] = 2;
	g.Tiles[2][2] = 16;
	g.Tiles[3][2] = 16;
	g.Move(DOWN);
	if !g.Equal(FromSparseGrid([]int{3, 0, 2, 2, 0, 8, 3, 2, 32}, 44)) {
		t.Fatalf("move down 1 failed")
	}
}

func TestPlaceRandom(t *testing.T) {
	g := NewGrid()

	if !g.PlaceRandom() {
		t.Fatalf("PlaceRandom on empty grid should return true")
	}
	var total GridNum = 0
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			total += g.Tiles[r][c]
		}
	}
	if total != 2 {
		t.Fatalf("PlaceRandom didn't place exactly one 2")
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			g.Tiles[r][c] = 4
		}
	}
	if g.PlaceRandom() {
		t.Fatalf("PlaceRandom on full grid should return false")
	}
}

File: grid_test.cpp
#include "grid.hpp"
#include <random>
#include <chrono>
#include <cassert>

void test_clone() {
	Grid g;
	g.grid[0][0] = 2;
    g.score = 10;

    Grid g2(g);

	assert(g == g2);

	g2.grid[0][1] = 20;
    assert(g != g2);
}

void test_movement() {
    Grid g;
	g.grid[0][0] = 2;
	g.grid[0][1] = 2;

    Grid comp = from_sparse_grid({0, 0, 2, 0, 1, 2}, 0);
    assert(g == comp);

	g.move(RIGHT);
    assert(g == from_sparse_grid({0, 3, 4}, 4));

	g.grid[1][2] = 4;
	g.move(UP);
    assert(g == from_sparse_grid({0, 2, 4, 0, 3, 4}, 4));

	g.move(LEFT);
    assert(g == from_sparse_grid({0, 0, 8}, 12));

	g.grid[1][0] = 2;
	g.grid[2][2] = 16;
	g.grid[3][2] = 16;
	g.move(DOWN);
    assert(g == from_sparse_grid({3, 0, 2, 2, 0, 8, 3, 2, 32}, 44));
}

void test_place_random() {
    Grid g;
    assert(g.place_random());

	gridnum total = 0;
	for (size_t r = 0; r < rows; r++) {
		for (size_t c = 0; c < cols; c++) {
			total += g.grid[r][c];
		}
	}
    assert(total == 2 || total == 4);


	for (size_t r = 0; r < rows; r++) {
		for (size_t c = 0; c < cols; c++) {
            g.grid[r][c] = 4;
		}
	}
    
    assert(!g.place_random());
}

int main() {
    // Seeds the random number generator
    gen.seed(std::chrono::system_clock::now().time_since_epoch().count());
    test_clone();
    test_movement();
    test_place_random();
}
